src/audio/host.c

```diff
@@ -350,6 +351,14 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
 	if (!hd->no_stream_position) {
 		hd->report_pos += bytes;
 
+		/* update position every period to support host query to
+		 * get the latest value if cont_update_posn == 1.
+		 */
+		if (hd->cont_update_posn == 1) {
+			pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+			mailbox_stream_write(dev->pipeline->posn_offset,
+					     &hd->posn, sizeof(hd->posn));
+		}
```

Comment:
I have tried multiple times to understand this PR, and I really don't get the idea.

If the idea is to send an IPC back to the host to update the timestamps every time a host DMA block is handled by the firmware, this is a VERY BAD IDEA.

if e.g. the host component retrieves blocks of 1ms, then you will have 1000 interrupts per second to update a timestamp, but the .pointer on the kernel side only needs this information on a period_elapsed event, or on a snd_pcm_status request.

It would be a much better idea to cache the timestamp in a memory window shared with the host and have the driver read from it.

Reply:
@plbossart I don't change IPC(interrupt) in this PR. I still keep the conditional check in the else if condition below, which is still depend on host_period_bytes(period size from HOST side,not  host component period in DSP firmware side), I only want to update host position data in host component period(1ms) but doensn't send IPC to kernel. Thanks.
```c
if (!hd->no_stream_position) {
	hd->report_pos += bytes;

+	if (hd->cont_update_posn) {  
+		/*
+		 ***** update position every period to support host query to get the
+		 * latest value if cont_update_posn == 1.****
+		 */
+		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+		mailbox_stream_write(dev->pipeline->posn_offset,
+				     &hd->posn, sizeof(hd->posn));
+	} 
       
       if (hd->host_period_bytes != 0 &&
		   hd->report_pos >= hd->host_period_bytes) {     
		/*
		 * host_period_bytes is set to zero to disable position update
		 * by IPC for FW version before 3.11, so send IPC message to
		 * driver according to this condition and report_pos.
		 */
		hd->report_pos = 0;
		/* send timestamped position to host
		* (updates position first, by calling ops.position())
		*/
		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
		mailbox_stream_write(dev->pipeline->posn_offset,
				     &hd->posn, sizeof(hd->posn));
		ipc_msg_send(hd->msg, &hd->posn, false);   <==Send IPC here, it depend on host_period_bytes
	}
}
```

Reply:
ok, if you update the position in a mailbox that's fine, but then why do you initiate an IPC in your .pointer function in the kernel? You are not using the mailbox information, are you? I really don't see how the two sides work together.


````c
snd_pcm_uframes_t mtk_adsp_pcm_pointer(struct snd_sof_dev *sdev,
				       struct snd_pcm_substream *substream)
{
	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
	struct snd_soc_component *scomp = sdev->component;
	struct snd_sof_pcm *spcm;
	struct sof_ipc_stream_posn posn;
	struct snd_sof_pcm_stream *stream;
	snd_pcm_uframes_t pos;
	int ret;

	spcm = snd_sof_find_spcm_dai(scomp, rtd);
	if (!spcm) {
		dev_warn_ratelimited(sdev->dev, "warn: can't find PCM with DAI ID %d\n",
				     rtd->dai_link->id);
		return 0;
	}

	stream = &spcm->stream[substream->stream];
	ret = snd_sof_ipc_msg_data(sdev, stream->substream, &posn, sizeof(posn));
	if (ret < 0) {
		dev_warn(sdev->dev, "failed to read stream position: %d\n", ret);
		return 0;
	}

	memcpy(&stream->posn, &posn, sizeof(posn));
	pos = spcm->stream[substream->stream].posn.host_posn;
	pos = bytes_to_frames(substream->runtime, pos);

	return pos;
}

````

Reply:
@plbossart yes, I write host position into mailbox information on DSP firmware, so I get the information in .pointer callback function mtk_adsp_pcm_pointer. Thanks.
```c
host.c
		mailbox_stream_write(dev->pipeline->posn_offset,
				     &hd->posn, sizeof(hd->posn));
```

Reply:
Ack @plbossart , the "ipc_msg_data" will read from the mailbox (not send an IPC) on all platforms. Most use the common implementation which is a call to snd_sof_dsp_mailbox_read().

So basicly this adds a new variable to the interface to configuration position updates:
  1) should position in mailbox be updated at a) DSP tick rate, b) host period rate or c) not at all
  2) should IPC be sent at a) DSP tick rate, b) host period rate, c) not at all

Sending IPC without mailbox update makes no sense, so 2 depends on 1.

On Intel platforms, we use 1c/2c  for production, but allow 1b+2b for debug.

@yaochunhung is seeing issues for production and would like the ability to configure 1a+2b (or 1a+2c ?) for mtk platforms.

Updating mailbox more often has a performance impact, but I think the SOF interface should allow this.

Reply:
@kv2019i  Thanks for detail description. For mediatek, we want to use 1a + 2b. Thanks.

Reply:
I am still confused, 'host_period_rate' means what? 
a) the ALSA period in the DDR ring buffer 
b) the size of the host buffer component in the firmware?

if this is b), then I maintain that 2b) is crazy.
if this is a), then I don't see how the precision is increased.

Reply:
@plbossart it is a), and we update mailbox data in b) when add the codes below,
```c
host.c in DSP firmware
+	if (hd->cont_update_posn) {  
+		/*
+		 ***** update position every period to support host query to get the
+		 * latest value if cont_update_posn == 1.****
+		 */
+		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+		mailbox_stream_write(dev->pipeline->posn_offset,
+				     &hd->posn, sizeof(hd->posn));
+	} 
```
then it will be gotten in .pointer callback when call snd_sof_ipc_msg_data function as @kv2019i  mentioned. thanks.

Reply:
@plbossart So it's the ALSA period_size. Whenever application calls snd_pcm_avail() (or related calls), this will do a hwsync, , and end up calling pointer() SOF DSP ops. For MTK, this would not use a MMIO or info from most recently sent position IPC, but rather read the position from mailbox directly (i.e. mapped DSP memory). Main drawback I see is the high amount of SRAM write traffic on DSP side, but other than that, this should get better accuracy.

Reply:
ok that makes sense, but now I am confused on what we do today before this PR? Is this saying that when a period_elapsed event is signaled, the firmware does not have the latest information? 

In other words, I wonder if this 'new' behavior should have been the default behavior from the start, and if we need the new flag in the IPC structure or just an unconditional change in the firmware?

Reply:
@plbossart As @kv2019i mentioned, there are two ways to update position information. one is (A). period_elapsed event and another is (B).application call snd_pcm_avail(or related calls). For case (A), there is no change with/without this PR. For case (B), firmware will have the latest information with this PR.  alsa_conformance_test will use both case (A) and (B).  I set a new flag "cont_update_posn" in IPC structure "sof_ipc_stream_params " so that user can decide to use this new behavior on their platforms or not. Thanks.